### PR TITLE
NH-3670 - Dynamic component should allow generic dictionary

### PR DESCRIPTION
**Pull request by rjperes**
https://nhibernate.jira.com/browse/NH-3670

Created additional tests with generic properties for NH-1796, NH-1039, NH-3571 and NH-2664

### CODE REVIEW

File: src/NHibernate/Tuple/DynamicMapInstantiator.cs [hunk 1]
```diff
@@ -48,7 +50,14 @@ public object Instantiate()
...
 		{
-			return new Hashtable();
+			if ((this.component != null) && (typeof(IDictionary<string, object>).IsAssignableFrom(this.component.Type.ReturnedClass) == true))
+			{
+				return new Dictionary<string, object>();
+			}
+			else
+			{
+				return new Hashtable();
+			}
```
Comment (fredericDelaporte) [inline src/NHibernate/Tuple/DynamicMapInstantiator.cs]: If we want to default to strong typed dictionary, we should invert the null logic here:

	if (this.component == null ||
		typeof(IDictionary<string, object>).IsAssignableFrom(this.component.Type.ReturnedClass)
	{
		return new Dictionary<string, object>();
	}
	else
	{
		return new Hashtable();
	}

File: src/NHibernate/Tuple/DynamicMapInstantiator.cs [hunk 2]
```diff
@@ -65,7 +74,7 @@ public bool IsInstance(object obj)
 			}
 			else
 			{
-				return false;
+				return obj is IDictionary<string, object>;
 			}
```
Comment (fredericDelaporte) [inline src/NHibernate/Tuple/DynamicMapInstantiator.cs]: If defaulting to strong typed dictionary, we would have to change the logic here too:

	var that = obj as IDictionary<string, object>;
	if (that != null)
	{
		if (entityName == null)
		{
			return true;
		}
		string type = (string)that[KEY];
		return type == null || isInstanceEntityNames.Contains(type);
	}
	else
	{
		return obj is IDictionary;
	}

**Comment by hazzik** [comment]
Rebased

File: src/NHibernate/Tuple/Component/DynamicMapComponentTuplizer.cs [hunk 3]
```diff
@@ -9,21 +13,48 @@ namespace NHibernate.Tuple.Component
...
+						owner = prop.PropertyType;
+					}
+
+					return typeof(IDictionary<string, object>).IsAssignableFrom(prop.PropertyType) ? typeof(IDictionary<string, object>) : typeof(IDictionary);
+				}
+				else
+				{
+					return typeof(IDictionary);
+				}
+			}
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Tuple/Component/DynamicMapComponentTuplizer.cs]: >What causes this simpler implementation not to be used here?

The `this.component.Type.ReturnedClass` property calls this property:

https://github.com/nhibernate/nhibernate-core/blob/2345bb47aa0f6a19ccd56230b284abed6002c9af/src/NHibernate/Type/ComponentType.cs#L99-L103

So, you cannot fetch the value of this property out of the void.

File: src/NHibernate/Tuple/DynamicMapInstantiator.cs [hunk 4]
```diff
@@ -48,7 +50,14 @@ public object Instantiate()
 
 		protected virtual IDictionary GenerateMap()
 		{
-			return new Hashtable();
+			if ((this.component != null) && (typeof(IDictionary<string, object>).IsAssignableFrom(this.component.Type.ReturnedClass) == true))
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Tuple/DynamicMapInstantiator.cs]: It's just his code style, however this needs to be changed to align with _our_ code style.

File: src/NHibernate/Tuple/DynamicMapInstantiator.cs (hunk 2 above)
Comment (hazzik) [reply src/NHibernate/Tuple/DynamicMapInstantiator.cs]: This piece code here only for `EntityMode.Map` instantiator.

File: src/NHibernate/Cfg/XmlHbmBinding/ClassBinder.cs [hunk 5]
```diff
@@ -301,16 +301,11 @@ protected string GetClassTableName(PersistentClass model, string mappedTableName
...
-				model.IsEmbedded = false;
-				model.IsDynamic = true;
-			}
-
 			else if (reflectedClass != null)
 			{
 				model.ComponentClass = reflectedClass;
 				model.IsEmbedded = false;
+				model.IsDynamic = componentMapping is HbmDynamicComponent;
 			}
```
Comment (hazzik) [inline src/NHibernate/Cfg/XmlHbmBinding/ClassBinder.cs]: Change to this file can be reverted if we do not care about "invalid" property declarations (see my [comment](https://github.com/nhibernate/nhibernate-core/pull/305/files#r193271703)).

File: src/NHibernate/Tuple/Component/DynamicComponentInstantiator.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace NHibernate.Tuple.Component
+{
+	[Serializable]
+	internal class DynamicComponentInstantiator : IInstantiator
+	{
+		private readonly bool _isGeneric;
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Tuple/Component/DynamicComponentInstantiator.cs]: @rjperes, @fredericDelaporte your thoughts?

[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Tuple/Component/DynamicComponentInstantiator.cs]: A possible breaking change, yes it is, but still a grey area one. The [documentation](http://nhibernate.info/doc/nhibernate-reference/mapping.html#mapping-declaration-component) clearly states to map it as an `IDictionary`.

> The `<dynamic-component>` element allows an `IDictionary` to be mapped as a component, where the property names refer to keys of the dictionary.

The fact it was a `HashTable` is an [implementation detail](http://nhibernate.info/doc/nhibernate-reference/persistent-classes.html#persistent-classes-tuplizers) on which the user should not rely:

> Users may also plug in their own tuplizers. Perhaps you require that a System.Collections.IDictionary implementation other than System.Collections.Hashtable be used while in the dynamic-map entity-mode;

I think we can accept this possible breaking change in a minor release. (With a small update to above example in documentation.)

[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Tuple/Component/DynamicComponentInstantiator.cs]: > > Users may also plug in their own tuplizers.

`<dynamic-component>` does not support custom tuplizers.

[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Tuple/Component/DynamicComponentInstantiator.cs]: I have removed the _isGeneric check as it did not work in all the cases. It seems that some code depends on `ComponentClass` being null for `<dynamic-component>`

[the remark this answers is not in the document]
Comment (fredericDelaporte) [reply src/NHibernate/Tuple/Component/DynamicComponentInstantiator.cs]: > `<dynamic-component>` does not support custom tuplizers.

Ok, I did not realize this PR was leaving dynamic entity mode "as-is", using `HashTable`, and was just changing the dynamic component. This creates some kind of inconsistency, although that still remains internal implementation details.

File: src/NHibernate/Linq/NhRelinqQueryParser.cs [hunk 7]
```diff
@@ -86,7 +87,7 @@ public NHibernateNodeTypeProvider()
 		public bool IsRegistered(MethodInfo method)
 		{
 			// Avoid Relinq turning IDictionary.Contains into ContainsResultOperator.  We do our own processing for that method.
-			if (method.DeclaringType == typeof(IDictionary) && method.Name == "Contains")
+			if ((method.DeclaringType == typeof(IDictionary) || method.DeclaringType == typeof(IDictionary<string, object>)) && method.Name == "Contains")
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Linq/NhRelinqQueryParser.cs]: Reverted.

File: src/NHibernate/Linq/Functions/DictionaryGenerator.cs [hunk 8]
```diff
@@ -58,7 +58,7 @@ public abstract class DictionaryRuntimeMethodHqlGeneratorBase<TGenerator> : IRun
 
 		public bool SupportsMethod(MethodInfo method)
 		{
-			return (method != null) && (method.Name == MethodName) && method.IsMethodOf(typeof(IDictionary));
+			return (method != null) && (method.Name == MethodName) && (method.IsMethodOf(typeof(IDictionary)) || method.IsMethodOf(typeof(IDictionary<string, object>)));
```
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Linq/Functions/DictionaryGenerator.cs]: Reverted.

File: src/NHibernate/Tuple/Component/DynamicComponentInstantiator.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hazzik) [reply src/NHibernate/Tuple/Component/DynamicComponentInstantiator.cs]: Yes, this PR was only about dynamic component.

**Comment by hazzik** [comment]
This is now ready for review.

**Comment by hazzik** [comment]
@fredericDelaporte I want to make dynamic component to support `dynamic` property, but this is for another PR as it requires much more work.

**Comment by hazzik** [comment]
I’ll merge myself

**Comment by hazzik** [comment]
I just realised that some of the interfaces has changed. I'll make a follow-up PR to remove the breaking changes.

**Pull request merged**